flocker/node/agents/ebs.py

```diff
@@ -152,6 +201,62 @@ def compute_instance_id(self):
         """
         return get_instance_metadata()['instance-id'].decode("ascii")
 
+    def _get(self, blockdevice_id):
+        """
+        Lookup BlockDeviceVolume for given blockdevice_id.
+
+        :param unicode blockdevice_id: ID of a blockdevice that needs lookup.
+
+        :returns BlockDeviceVolume for the given input id.
+        """
+        for volume in self.list_volumes():
+            if volume.blockdevice_id == blockdevice_id:
+                return volume
+        raise UnknownVolume(blockdevice_id)
+
+    def _get_ebs_volume(self, blockdevice_id):
+        """
+        Lookup EBS Volume information for a given blockdevice_id.
+
+        :param unicode blockdevice_id: ID of a blockdevice that needs lookup.
+
+        :returns boto.ec2.volume.Volume for the input id. ``None`` if
+            no boto.ec2.volume.Volume was found for the given id.
+        """
+        for volume in self.connection.get_all_volumes():
+            if volume.id == blockdevice_id:
+                # Sync volume for uptodate metadata
+                volume.update()
+                return volume
+        return None
+
+    def _next_device(self, instance_id):
+        """
+        Get the next available EBS device name for a given EC2 instance.
+        Algorithm:
+        1. Get all ``Block devices`` currently in use by given instance:
+            a) List all volumes visible to this instance.
+            b) Gather device IDs of all devices attached to (a).
+        2. Devices available for EBS volume usage are ``/dev/sd[f-p]``.
+           Find the first device from this set that is currently not
+           in use.
+
+        :param unicode instance_id: EC2 instance ID.
+
+        :returns unicode file_name: available device name for attaching
+            EBS volume.
+        :returns ``None`` if suitable EBS device names on this EC2
+            instance are currently occupied.
+        """
+        volumes = self.connection.get_all_volumes()
+        devices = [v.attach_data.device for v in volumes
+                   if v.attach_data.instance_id == instance_id]
```

Comment:
Sorry for not being clear in the comments. `attach_data.device` points to the device name you chose during attach_volume. The tag i added carries the eventual device name picked by OS's device driver. For example `attach_volume(..., '/dev/sdg',...)` might end up attaching the volume to `/dev/xvdh`. All EC2 instance APIs, including `attach_data.device` will show `/dev/sdg`. The tag i add for the volume will show `/dev/xvdh`. `get_device_path` needs `/dev/xvdh` because that is what the OS recognizes, so, we need to use that for creating file systems, mounting, etc. `/dev/sdg` is pretty much useless for our consumption.